### PR TITLE
merge AndAndExp and OrOrExp into LogLogExp

### PR DESCRIPTION
**Pull request by WalterBright**
The two are very nearly identical, so merge them and eliminate code.

### CODE REVIEW

**Comment by ibuclaw** [comment]
What is a LogLog expression?

**Comment by WalterBright** [comment]
Logical Logical operators

**Comment by ibuclaw** [comment]
Maybe it should just be called either one of LogExp (logical expression), LogicalExp, or BinLogExp (binary logical expression).

Reading out aloud the name logical-logical operator gives me the same feeling of incorrectness that that-that or had-had would. You have to read it twice to make sure you understand it.

File: src/ddmd/astbase.d [hunk 1]
```diff
@@ -5560,24 +5560,11 @@ struct ASTBase
         }
     }
 
-    extern (C++) final class AndAndExp : BinExp
+    extern (C++) final class LogLogExp : BinExp
```
Comment (JinShil) [inline src/ddmd/astbase.d]: Could use a comment explaining that `LogLogExp` (or whatever name is ultimately decided on) is for `&&` _and_ `||` expressions so contributors aren't left guessing.

**Comment by JinShil** [comment]
As a new spelunker of DMD source code, I have to say I'm not crazy about this change; it hides something that was previously quite obvious behind something more obscure.  

If the `AndAndExp` and `OrOrExp` share common functionality, then to whatever extent it is possible, create an `abstract LogExp` class and concrete `AndAndExp` and `OrOrExp` classes, and try to refactor the common code to the base `LogExp`.  `visit(LogExp)` can contain the common functionality with `visit(AndAndExp)` and `visit(OrOrExp)` forwarding to it.

**Comment by JinShil** [comment]
> As a new spelunker of DMD source code, I have to say I'm not crazy about this change

...but the elimination of redundant code is nice.

**Comment by WalterBright** [comment]
I've used `loglog` for the andand/oror expressions for a very long time:
https://github.com/dlang/dmd/blob/master/src/ddmd/backend/cod2.c?utf8=%E2%9C%93#L2223

**Comment by ibuclaw** [comment]
Well this being shared code across potentially many apps, there is a need for it to be understood by anyone reading the api docs (if/when they get published).

File: src/ddmd/astbase.d (hunk 1 above)
Comment (WalterBright) [inline src/ddmd/astbase.d]: That's https://github.com/dlang/dmd/pull/7061

**Comment by UplinkCoder** [comment]
@WalterBright I agree with @JinShil here.
LogLogExp is a headscratcher.
couldln't you introduce it as a superclass of AndAndExp and OrOrExp ?

**Comment by ibuclaw** [comment]
> couldln't you introduce it as a superclass of AndAndExp and OrOrExp ?

Hmm, right, I've just noticed that this removes AndAnd and OrOrExp declarations, this breaks downstream projects that use these.

At the very least keep the original classes around, update c++ headers, then this is OK for inclusion.

**Comment by WalterBright** [comment]
> couldln't you introduce it as a superclass of AndAndExp and OrOrExp ?

That wouldn't eliminated near as much code. What I did here is hardly different from IdentityExp and CmpExp, to name just two.

> this breaks downstream projects that use these.

Indeed it does. The fix is generally to delete code (most of this PR is just deleting code).

> At the very least keep the original classes around

Why?

> update c++ headers

expression.h is already updated in this PR

**Comment by jacob-carlborg** [comment]
> Maybe it should just be called either one of LogExp (logical expression), LogicalExp, or BinLogExp (binary logical expression).

Yes, please. Ideally it should be called `LogicalExpression`, but to be consistent with the rest of the code, `LogicalExp` would be acceptable as well.

Keep in mind that we want to expose DMD as a library, then we need to think about the names we pick and avoid abbreviations.

**Comment by ibuclaw** [comment]
> Why?

There's no cost keeping them around, and downstream can choose to update independently in their own time, if they even choose to.

**Comment by WalterBright** [comment]
renamed LogLogExp to LogicalExp

**Review by wilzbach — APPROVED** [review]
> +57 / -300

This is amazing!

**Comment by wilzbach** [comment]
> There's no cost keeping them around

There is always a hidden cost - especially when it's about maintaining deprecations.

> and downstream can choose to update independently in their own time, if they even choose to.

AFAICT DMD has never issued deprecations before and as DaaL (DMD as a library) isn't really used yet or rarely used, I don't see why this should start with this PR.
Besides the change itself here is really trivial from two classes into one.

**Comment by JinShil** [comment]
This change is fine.  The reduction in code is nice, and as I spend more time with the DMD code, I realize it is idiomatic DMD.  However, it may be worth remembering that DMD was translated from C++ (whose idioms were more C-ish than C++-ish), and this change seems to favor the C-ish idioms over the modeling power of D.

That being said, the `LogicalExp` name is better and the reduction in code is probably worth it, but those benefits are being traded for a less intuitive API, IMO.

File: src/ddmd/sideeffect.d [hunk 2]
```diff
@@ -287,14 +287,10 @@ extern (C++) bool discardValue(Expression e)
...
     case TOKandand:
-        {
-            AndAndExp aae = cast(AndAndExp)e;
-            return discardValue(aae.e2);
-        }
     case TOKoror:
         {
-            OrOrExp ooe = cast(OrOrExp)e;
-            return discardValue(ooe.e2);
+            LogicalExp aae = cast(LogicalExp)e;
```
Comment (JinShil) [inline src/ddmd/sideeffect.d]: Nit: `aae` probably isn't such a good name anymore.

File: src/ddmd/e2ir.d [hunk 3]
```diff
@@ -3373,13 +3373,13 @@ elem *toElem(Expression e, IRState *irs)
         /***************************************
          */
 
-        override void visit(AndAndExp aae)
+        override void visit(LogicalExp aae)
```
Comment (JinShil) [inline src/ddmd/e2ir.d]: Nit: `aae` probably isn't such a good name anymore.

File: src/ddmd/staticcond.d [hunk 4]
```diff
@@ -40,25 +40,23 @@ import ddmd.utils;
 
 bool evalStaticCondition(Scope* sc, Expression exp, Expression e, ref bool errors)
 {
-    if (e.op == TOKandand)
+    if (e.op == TOKandand || e.op == TOKoror)
     {
-        AndAndExp aae = cast(AndAndExp)e;
+        LogicalExp aae = cast(LogicalExp)e;
```
Comment (JinShil) [inline src/ddmd/staticcond.d]: Nit: `aae` probably isn't such a good name anymore.

**Review by JinShil — CHANGES_REQUESTED** [review]

**Comment by jacob-carlborg** [comment]
> renamed LogLogExp to LogicalExp

Thanks.

**Comment by andralex** [comment]
@WalterBright thanks very much for compromising on the naming, it certainly threw me off as much as some of the folks who chimed in in here.

**Review by andralex — APPROVED** [review]

**Pull request merged**